### PR TITLE
NOISSUE - Add twins related documentation page

### PR DESCRIPTION
**Pull request by darkodraskovic**
Signed-off-by: Darko Draskovic <darko.draskovic@gmail.com>

### CODE REVIEW

File: docs/twins.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+### CRUD operations
+
+
+
+#### Create && Update
+
+Create and update request
+
+```
```
Comment (mteodor) [inline docs/twins.md:190]: add ```json

File: docs/twins.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+role, position, function in the overall system - into their **metadata**.
+Although this might work - with a lot of additional effort of writing the
+relatively complex code to parse things' and channels' metadata - it is not a
+practical approach and we still don't get - at least not out of the box - a nice
+overview of the system as a whole and of its components.
+
+To overcome these two problems, Mainflux comes with a **digital twin service**.
+Mainflux digital twin consists of three parts:
+
+- general data about twin itself, i.e. **twin's metadata**,
```
Comment (drasko) [inline docs/twins.md]: I would capitalize bullets, as you did below (for consistency)

File: docs/twins.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        "persist_state": true
+      }
+    ],
+    "delta": 1
+  }
+}
+```
+
+
+```
```
Comment (drasko) [inline docs/twins.md]: Add `bash` after triple backtick

**Review by drasko — CHANGES_REQUESTED** [review]
Generally this looks well, but I would appreciate a diagram that shows where Digital Twin service sits in the system (connected to NATS to get the requests, sends notifications upon deltas, exposes API - so connected to NginX as well, and so on).

File: docs/twins.md [hunk 4]
```diff
@@ -65,77 +77,97 @@ Attributes has a state persistance flag that determines whether the messages
 communicated by it's corresponding channel and subtopic trigger the creation of
 a new twin state.
 
-A JSON representation of twin might look like this:
+A JSON representation of twin might look like this when we define our digital
+twin:
 
 ```
```
Comment (manuio) [inline docs/twins.md]:
```
``` json 
```

File: docs/twins.md [hunk 5]
```diff
@@ -144,10 +176,76 @@ corresponds to the attribute name and a value is the actual value of the
 attribute. All [SenML value
 types](https://tools.ietf.org/html/rfc8428#section-4.3) are supported.
 
+A JSON representation of a partial list of states might look like this:
 
-### CRUD operations
+```
```
Comment (manuio) [inline docs/twins.md]:
```
``` json 
```

File: docs/twins.md [hunk 6]
```diff
@@ -211,10 +309,18 @@ curl -s -S -i -X GET -H "Authorization: <user_auth_token>" http://localhost:8191
 curl -s -S -i -X DELETE -H "Authorization: <user_auth_token>" http://localhost:8191/twins/<twin_id>
 ```
 
+
 ### STATES operations
 
 #### List
 
 ```
```
Comment (manuio) [inline docs/twins.md]:
```
``` bash 
```

**Review by manuio — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
I think `WIP` label can be removed

**Comment by drasko** [comment]
Also, please edit PR title and align it to Mainflux standard by prefixing it with `NOISSUE - `

**Review by drasko — DISMISSED** [review]
LGTM

File: docs/twins.md [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+For example, an industrial machine can use multiple protocols such as MQTT,
+OPC-UA, a regularly updated machine hosted CSV file etc. to send measurement
+data - such as flowrate, material temperature, etc. - and state metadata - such
+as engine and chassis temperature, engine rotations per seconds, identity of the
+current human operator, etc. - as well as to receive control, i.e. actuation
+messages - such as, turn on/off light, increment/decrement borer speed, switch
+knifes, etc.
+
+Digital twin is an abstract - and usually less detailed - digital replica of a
```
Comment (dborovcanin) [inline docs/twins.md]: ... - a digital replica of a real-world...

File: docs/twins.md [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+real world system such as the industrial machine we have just described. It is
+used to create and store information about system's state at any given moment,
+to compare system state over a given period of time - so called diffs or
```
Comment (dborovcanin) [inline docs/twins.md]: ... so-called...

File: docs/twins.md [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+deltas - as well as to control agents composing the system.
+
+### What is Mainflux digital twin
+
+Any data producer or data consumer - which we refer to here collectively as data
+agent - or an interrelated system of data agents, can be represented by means of
+possibly multiple [Mainflux things, channels and subtopics]((architecture.md)).
```
Comment (dborovcanin) [inline docs/twins.md]: ..., and subtopics...

File: docs/twins.md [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+For example, an OPC-UA server can be represented as a Mainflux thing and its
+nodes can be represented as multiple Mainflux channels or multiple subtopics of
+a single Mainflux channel. What is more, you can invert the representation: you
+can represent server as a channel and node as things. Mainflux platform is meant
+to empower you with the freedom of expression so you can make a digital
+representation of any data agent according to your needs.
+
+Although this works well, satisfies the requirements of a wide variety of use
+cases and corresponds to the intended use of Mainlfux IoT platform, this setup
```
Comment (dborovcanin) [inline docs/twins.md]: ..., and corresponds...

File: docs/twins.md [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+twins. Mainflux digital twin consists of three parts:
+
+- General data about twin itself, i.e. **twin's metadata**,
+- History of twin's **definitions**, including current definition,
+- History of twin's **states**, including current state.
+
+Before we dwell into twin's anatomy, it is important to realize that in order to
+use Mainflux twin service, you have to [provision Mainflux things and
+channels](provisioning.md) and you have to connect things and channels
+beforehand. As you go, you can modify your things, channels and connections and
```
Comment (dborovcanin) [inline docs/twins.md]: ..., and connections...

File: docs/twins.md [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+by Mainflux user -, twin's ID (unique) and name (not necessarily unique), twin's
+creation and update dates as well as twin's revision number. The latter refers
+to the sequential number of twin's definition.
+
+The twin's **definition** is meant to be a semantic representation of system's
+data sources and consumers (data agents). Each data data agent is represented by
+means of **attribute**. Attribute consists of data agent's name, Mainflux
+channel and subtopic over which it communicates. Nota bene: each attribute is
+uniquely defined by the combination of channel and subtopic and we cannot have
+two or more attributes with a same channel and subtopic in the same definition .
```
Comment (dborovcanin) [inline docs/twins.md]: ... the same channel and subtopic...

File: docs/twins.md [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+Attributes has a state persistance flag that determines whether the messages
```
Comment (dborovcanin) [inline docs/twins.md]: Attributes have a state persistence...

File: docs/twins.md [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+communicated by it's corresponding channel and subtopic trigger the creation of
```
Comment (dborovcanin) [inline docs/twins.md]: ..its corresponding channel...

File: docs/twins.md [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+a new twin state. Twin states are persisted in the separate collection of the
+same database. Currently, twins service uses the MongoDB. InfluxDB support for
+twins and states persistance is on the roadmap.
```
Comment (dborovcanin) [inline docs/twins.md]: ...persistence...

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: docs/twins.md [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+            "delta": 1000000
+        }
+    ]
+}
+```
+
+In the case of the upper twin, we begin with an empty definition, the one with
+the `id` **0** - we could have provided the definition immediately - and over
+the course of time, we add two more definitions, so the total number of
+revisions is **2** (revision index is zero based). We decide not to persist the
```
Comment (dborovcanin) [inline docs/twins.md]: ...zero-based...

File: docs/twins.md [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+system. `delta` is the number of nanoseconds used to determine whether the
+received attribute value should trigger the generation of the new state or the
+same state should be updated. The reason for this is to enable state sampling
+over the regular intervals of time. Discarded values are written to the database
+of choice by Mainflux [writers](storage.md), so you can always retrieve
+intermediate values if need be.
+
+
+**states** are created according to the twin's current definition. A state
+stores twin's ID - every state belongs to a single twin -, its own ID, twin's
```
Comment (dborovcanin) [inline docs/twins.md]: typo `,` in  `...twin -, its...`

File: docs/twins.md [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+definition number, creation date and the actual payload. **Payload** is a set of
```
Comment (dborovcanin) [inline docs/twins.md]: ..., and the actual...

File: docs/twins.md [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+```
+
+As you can see, the first two states correspond to the definition **1** and have
+only two attributes in the payload. The rest of the states is based on the
+definition **2**, where we persist three attributes and, as a consequence, its
+payload consists of three entries.
+
+### Authentication and authorization
+
+Twin belongs to a Mainflux user, tenant representing physical person or
```
Comment (dborovcanin) [inline docs/twins.md]: ...representing a physical person...

File: docs/twins.md [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+organization. User owns Mainflux things and channels as well as twins. Mainflux
+user provides authorization and authentication mechanisms to twins service. For
+more details, please see [Authentication with Mainflux keys](authentication.md).
+In practical terms, we need to create a Mainflux user in order to create digital
```
Comment (dborovcanin) [inline docs/twins.md]: ...in order to create a digital twin.

File: docs/twins.md [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+
+### TWIN operations
+
+For the full and up to date Mainflux twins service HTTP API please refer to the
+[twins service swagger
+file](https://github.com/mainflux/mainflux/blob/master/twins/swagger.yaml).
+
+#### Create && Update
+
+Create and update request use JSON body to initialize and modify, respectively,
```
Comment (dborovcanin) [inline docs/twins.md]: Create and update requests... (plural)

File: docs/twins.md [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+
+In order to pick up this notifications, you have to create a Mainflux channel
+before you start the twins service and inform the twins service about the
+channel by means of the environment variable, like this:
+
+```bash
+export MF_TWINS_CHANNEL_ID=f6894dfe-a7c9-4eef-a614-637ebeea5b4c
+```
+
+The twins service will use the this channel to publish notifications related to
```
Comment (dborovcanin) [inline docs/twins.md]: ...will use ~the~ this channel...

File: docs/twins.md [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+- `save.failure` - on state save failure.
+
+Normally, you can use NATS wildcards. In order to learn more about Mainflux
+channel topic composition, please read about [subtopics](messaging.md). The
+point is to be able to subscribe to all subjects or any operation pair subject -
+e.g. create.success/failure - by means of one connection and read all messages
+or all operation related messages in the context of the same subscription.
+
+Since messages published on NATS are republished on any other protocol supported
+by Mainflux - HTTP, MQTT, CoAP and WS - you can use any supported protocol
```
Comment (dborovcanin) [inline docs/twins.md]: ..., and WS...

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
In the architecture diagram, Things and Users services point to the same LocalDB. Even though it has nothing to do with Twins service, this can be misleading because these services use separate DBs. Can you use separate databases or remove `Local DB` from the diagram. If you decide to keep it, please align the color legend(red for storage) with it.

File: docs/twins.md (hunk 7 above)
Comment (darkodraskovic) [reply docs/twins.md]: There is an `a` in the whole phrase `an abstract - and usually less detailed - digital replica`.

File: docs/twins.md (hunk 17 above)
Comment (darkodraskovic) [reply docs/twins.md]: You can put a comma after `-` if you target that.

File: docs/twins.md [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+
+### What is digital twin
+
+**Twin** refers to a **digital representation** of a **real world data system** consisting of possibly multiple data sources/producers and/or destinations/consumers (data agents).
+
+For example, an industrial machine can use multiple protocols such as MQTT, OPC-UA, a regularly updated machine hosted CSV file etc. to send measurement data - such as flowrate, material temperature, etc. - and state metadata - such as engine and chassis temperature, engine rotations per seconds, identity of the current human operator, etc. - as well as to receive control, i.e. actuation messages - such as, turn on/off light, increment/decrement borer speed, change blades, etc.
+
+Digital twin is an abstract - and usually less detailed - digital replica of a real world system such as the industrial machine we have just described. It is used to create and store information about system's state at any given moment, to compare system state over a given period of time - so-called diffs or deltas - as well as to control agents composing the system. 
+### Mainflux digital twin
```
Comment (manuio) [inline docs/twins.md:13]: You could add a new line before this tittle

File: docs/twins.md [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+```json
+{
+    "id": 0,
+    "created": "2020-05-05T08:41:39.142Z",
+    "attributes": [],
+    "delta": 1000000
+}
+```
+
+will be created.
```
Comment (manuio) [inline docs/twins.md:254]: Is this ok?

**Review by dborovcanin — APPROVED** [review]

File: docs/twins.md (hunk 25 above)
Comment (darkodraskovic) [reply docs/twins.md:254]: Yes, cause I consistently make a newline after the code block. See here how it looks compiled https://github.com/darkodraskovic/docs/blob/twins/docs/twins.md#create

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**